File: de.tototec.tools.cmvn/src/main/java/de/tototec/tools/cmvn/configfile/KeyValueWithOptions.java
package de.tototec.tools.cmvn.configfile;

import java.util.LinkedList;
import java.util.List;

import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@EqualsAndHashCode(callSuper = true)
public class KeyValueWithOptions extends KeyValue {
	private final List<KeyValue> options;

	public KeyValueWithOptions(final String key, final String value,
			final List<KeyValue> options) {
		super(key, value);
		this.options = options;
	}

	public KeyValueWithOptions(final String key, final String value,
			final String optionSeparator, final String keyValueSeparator,
			final String defaultValue) {
		super(key, value.split(optionSeparator)[0].trim());

		final StringSplitter splitter = new StringSplitter();

		options = new LinkedList<KeyValue>();

		final String[] split = splitter.split(value, optionSeparator, "\\");
		if (split.length > 1) {
			for (int i = 1; i < split.length; ++i) {
				final String option[] = splitter.split(split[i],
						keyValueSeparator, "\\", 2);
				final String oValue = option.length == 1 ? defaultValue
						: option[1].trim();
				options.add(new KeyValue(option[0].trim(), oValue));
			}
		}
	}

	public KeyValueWithOptions(final KeyValue keyValue,
			final String optionSeparator, final String keyValueSeparator,
			final String defaultValue) {
		this(keyValue.getKey(), keyValue.getValue(), optionSeparator,
				keyValueSeparator, defaultValue);
		setFile(keyValue.getFile());
		setLine(keyValue.getLine());
	}

}

File: de.tototec.cmvn/src/main/scala/de/tototec/cmvn/configfile/KeyValueWithOptions.scala
package de.tototec.cmvn.configfile

import java.util.LinkedList
import java.util.List
//remove if not needed
import scala.collection.JavaConversions._

class KeyValueWithOptions private (key: String, value: String)
  extends KeyValue(key, value) {

  private var _options: List[KeyValue] = _
  def options = _options

  def this(key: String, value: String, options: List[KeyValue]) {
    this(key, value)
    this._options = options
  }

  def this(key: String,
           value: String,
           optionSeparator: String,
           keyValueSeparator: String,
           defaultValue: String) {
    this(key, value.split(optionSeparator)(0).trim())
    val splitter = new StringSplitter()
    this._options = new LinkedList[KeyValue]()
    val split = splitter.split(value, optionSeparator, "\\")
    if (split.length > 1) {
      for (i <- 1 until split.length) {
        val option = splitter.split(split(i), keyValueSeparator, "\\", 2)
        val oValue = if (option.length == 1) defaultValue else option(1).trim()
        options.add(new KeyValue(option(0).trim(), oValue))
      }
    }
  }

  def this(keyValue: KeyValue,
           optionSeparator: String,
           keyValueSeparator: String,
           defaultValue: String) {
    this(keyValue.key, keyValue.value, optionSeparator, keyValueSeparator, defaultValue)
    this.file = keyValue.file
    this.line = keyValue.line
    for (option <- options) {
      option.file = keyValue.file
      option.line = keyValue.line
    }
  }
}
